instrument_server update

netshade opened the pull request:
Some long required love for `instrument_server`. Summary of changes:

* Remove `json` dependency, which should lower the number of server dependencies required to install the gem (`json` inadvertently requires the `ruby-dev` package on Ubuntu, which installs the `build-essential` package, which seems like overkill for what we're doing here)
* Remove the requirement that an API key be provided when issuing `instrument_server restart` or `instrument_server stop`.
* Allow configuring the reporting interval, and change the default interval to 30s.
* Change the default location of the PID and log files to `$HOME/the_file`. The previous location, `/tmp/`, could cause the `PID` to be cleaned out from underneath the process.
* Allow the PID location and log file location to be specified at startup.
* Try to ensure the process has started successfully before detaching, and fail loudly if it has not.
* Output the number of metrics sent at each reporting interval.
* Add a slim framework for writing custom metric scripts ([CUSTOM_METRICS.md](https://github.com/expectedbehavior/instrumental_tools/blob/increase-process-durability/CUSTOM_METRICS.md)), as well as some initial examples of how to use it. Documentation as well.
* Set the required Ruby version to 1.9.
* Monolithic `instrument_server` script broken apart into several files to ease w/ maintenance
* Add scripts for building `rpm`, `tgz` and `deb` packages that allow a sysadmin to install instrument_server w/ zero dependencies on external packages
* Allow process privilege to be dropped to a lesser privileged user at init
* Add functionality to push packages to [packagecloud.io](http://packagecloud.io) for redistribution
* Add several example init scripts for starting `instrument_server` at system boot time
* Removes dependencies on `cat`, `grep`, `mount` and `sysctl` to extract Linux host statistics
* Add example `chef` recipe and `puppet` module for installing on servers
* Add KitchenCI to test installation procedure and verify running 

Release TODO:

- [x] Update all internal refs to release numbers before release
- [x] Decide on publicly available `tgz` download links in order for documentation and scripts to refer to same.

janxious commented on CUSTOM_METRICS.md:
```diff
@@ -0,0 +1,56 @@
+
+A custom script may be a binary or shell script that exists in the custom scripts directory (`$HOME/.instrumental_scripts`). Each time the `instrument_server` process collects system metrics, it will also execute your script with the following arguments:
+
+* Argument 1: The Unix timestamp of the last time this script had been executed, in seconds. If the process has not successfully been ran by `instrument_server` before, this value will be 0.
```
"been run"

spacetc62 commented:
why $HOME/the_file and not /var/run/blahblah?

janxious commented on README.md:
````diff
@@ -25,15 +25,22 @@ instrument_server -k <API_KEY> -H <HOSTNAME>
 
 ```
-instrument_server -k <API_KEY> -d
+instrument_server -k <API_KEY> start
````
It's stupid, but this is a breaking change and maybe worth a major version change?

janxious replied:
P.S. pretty exciting changes.

netshade replied:
Yeah, I could definitely see that.

netshade commented:
RE: why `$HOME` and not `/var/run`, in some cases `/var/run/` can be owned by a higher privileged user, and I don't want to encourage escalating this processes' privileges if it's going to be executing scripts on that user's behalf. I figured the user can always set it to `/var/run` if they have a pref.

spacetc62 commented on lib/instrumental_tools/metric_script_executor.rb:
```diff
@@ -0,0 +1,96 @@
+
+  def run
+    process_to_output = {}
+    if can_execute_in_directory?(directory)
```
oh my god such a long if statement - break this method into some smaller methods?

netshade replied:
It's pretty long, but as the `main` method, it seems excusable for what it's doing. I'd prefer to keep it as is in form.

netshade replied:
NOTE: Discussed in person, @spacetc62 beat me with a rubber hose until I relented on moving this.

spacetc62 commented on lib/instrumental_tools/metric_script_executor.rb:
```diff
@@ -0,0 +1,96 @@
+                              :err   => stderr_w)
+
+          exit_status = nil
+          exec_time   = Benchmark.realtime do
```
do you want to give a max time on these guys?  maybe even execute them in threads so that you can still be assured once every interval even if there are a lot?

netshade replied:
I'd prefer not to - timeouts should be the executed script's responsibility, we won't know what they're doing / touching.

janxious commented on lib/instrumental_tools/system_inspector.rb:
```diff
@@ -0,0 +1,67 @@
+      end
+  end
+
+  def self.command_missing(command, section)
```
`command_missing?`?

netshade replied:
Just to handle the case where a command that the script depends on is not present - no relation to `method_missing`.

janxious replied:
Yeah, I misread this entirely. I was thinking it was a boolean check because <reason I no longer remember>

janxious commented on lib/instrumental_tools/system_inspector/linux.rb:
```diff
@@ -0,0 +1,145 @@
+    def self.memory
+      _, total, used, free, shared, buffers, cached = `free -k -o | grep Mem`.chomp.split(/\s+/)
+      {
+        'memory.used_mb' => used.to_f / 1024,
```
Time to monkey patch. O_O

No, but maybe a method for `to_megabytes` or something that these bit all use.

Or maybe just a constant.

netshade replied:
Yeah, that'd make sense. Slim these up a bit.

janxious commented on lib/instrumental_tools/system_inspector/osx.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def self.netstat(interface = 'en1')
+      # mostly functional network io stats
+      headers, *lines = `netstat -ibI #{interface}`.split("\n").map { |l| l.split(/\s+/) } # FIXME: vulnerability?
```
What vuln are you worried about here?

netshade replied:
I can't say I recall - maybe @jqr remembers?

spacetc62 commented on examples/mysql/mysql_status.rb:
```diff
@@ -0,0 +1,78 @@
+
+previous_run_time = ARGV[0].to_i
+current_time      = Time.now.to_i
+run_interval      = (current_time - previous_run_time).to_f
```
so part of what comes back from mysql is an uptime variable.  You can use that for two things - 1) the canary metric, if uptime is less than last see uptime, and 2) for more correct calculation of time between current info and previous info

janxious replied:
@spacetc62 So are you proposing like a notice here "mysql died" or something?

spacetc62 commented on lib/instrumental_tools/server_controller.rb:
```diff
@@ -0,0 +1,62 @@
+        count += 1
+      end
+      custom_metrics.run.each do |(stat, value, time)|
+        agent.gauge("#{options[:hostname]}.#{stat}", value, time)
```
how do I do an increment and not a gauge with custom metrics?

netshade replied:
You don't - we should document that fact tho. I had trouble thinking of obvious cases where we'd care to support that.

spacetc62 replied:
seems like it would pretty reasonable to prefix each stat string with "gauge" or "increment" - and now would be the time to do that as we are defining the script input/output

netshade replied:
I'd be down for suffix, but not prefix. I would prefer that, if it had the data the user wanted, they could write a script that just did a `cat` of a procfs file, which would map directly to our format here.

spacetc62 commented on examples/mysql/mysql_status.rb:
```diff
@@ -0,0 +1,78 @@
+MYSQL_DEFAULTS_FILE     = ENV["MYSQL_DEFAULTS_FILE"]
+MYSQL_PASSWORD          = ENV["MYSQL_PASSWORD"]
+
+RATE_METRICS_TO_INSPECT = %w{Queries Bytes_sent Bytes_received Connections Slow_queries}
```
can we downcase these guys in the actual metric names we return to instrumental?

netshade replied:
Sure, will do.

netshade commented:
From @spacetc62: The `shellwords` escaping can be omitted in favor of `Process.spawn`'s arguments passing interface.

janxious commented on lib/instrumental_tools/metric_script_executor.rb:
```diff
@@ -0,0 +1,96 @@
+    end
+    process_to_output.flat_map do |path, (status, time, output)|
+      if status && status.success?
+        prefix = File.basename(path).split(".")[0..-2].join(".").gsub(/[^a-z0-9\-\_\.]/i, "_")
```
probably not a real danger but unicode

netshade replied:
In the metric name portion or the split? I'm not sure which transform of this you're referring to here

janxious replied:
The split should be fine. If we don't support unicode in metric names, then this is cool. I don't remember, honestly.

netshade replied:
Ah, I see - yeah, this should mirror the agent's validation at least. Willfix.

spacetc62 commented on examples/mysql/mysql_status.rb:
```diff
@@ -0,0 +1,78 @@
+# Ensure your instrument_server process runs with these environment variables set,
+# or replace with your own.
+
+MYSQL_HOST              = ENV["MYSQL_HOST"]
```
I think you can default to localhost here?

netshade replied:
Good point, will do

spacetc62 commented on examples/mysql/mysql_status.rb:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env ruby
```
how would you recommend running this same script multiple times against multiple different mysql servers?

netshade replied:
Create `n` copies of this file, each of which is intended to capture metrics for a specific server. Change the constants at the head of each file to reflect that specific server's information.

spacetc62 commented on lib/instrumental_tools/metric_script_executor.rb:
```diff
@@ -0,0 +1,96 @@
+    process_to_output.flat_map do |path, (status, time, output)|
+      if status && status.success?
+        prefix = File.basename(path).split(".")[0..-2].join(".").gsub(/[^a-z0-9\-\_\.]/i, "_")
+        output.split(/[\r\n]+/)                                                                # each line
```
why use split and not the each_line method?

netshade replied:
I forgot it existed, will change.

spacetc62 commented:
might want to have a non-ruby example script, just to show how that would feel?

spacetc62 commented on CUSTOM_METRICS.md:
````diff
@@ -0,0 +1,56 @@
+Your script is expected to output data in the following format on `STDOUT` in order to be sent back to Instrumental:
+
+```
+METRIC_NAME METRIC_VALUE
````
should add docs about optional time param

netshade replied:
Done

spacetc62 commented on CUSTOM_METRICS.md:
```diff
@@ -0,0 +1,56 @@
+
+Additionally, all scripts to be ran by the `instrument_server` process must be readable/writable only by the user the process is executing as (`0700`).
+
+## <a name="writing_custom_scripts"></a> Writing Custom Scripts
```
weird link here

netshade replied:
It's an anchor for internal doc link.

spacetc62 commented on lib/instrumental_tools/metric_script_executor.rb:
```diff
@@ -0,0 +1,96 @@
+        output.split(/[\r\n]+/)                                                                # each line
+          .map    { |line| line.split(/\s+/) }                                                 # split by whitespace
+          .select { |data| (2..3).include?(data.size)  }                                       # and only valid name value time? pairs
+          .map    { |(name, value, specific_time)| [[prefix, name].join("."), value.to_f, specific_time || time] } # with value coerced to a float
```
do you need to "to_i" the value from specific_time if there is a value?

netshade replied:
The agent should coerce it, but I don't like passing up a String or a Fixnum to the caller, would much prefer that the contract is always a Fixnum.

netshade commented:
Reading more about the way to accomplish things like net counters in Docker, it basically means we have to run these scripts as `root`. I can't say I'm a big fan of ever telling a user to run potentially untrusted scripts as root, I really want this process to execute as a lower privileged user if at all possible.

I'm going to modify the `docker` script to give instructions to the user on what to do as `root` to make this happen, I think that's the most reasonable compromise for now.

netshade commented:
Lots of updates. Along w/ the gem, we also can ship a tarball or package that has no external dependencies. When you install the package, it also will install itself into system startup. Along with these changes, we now also support loading the config in via `/etc/instrumental.yml`.

Since we can now come in via system startup, I've also modified the script to drop privileges to a safer user than `root` when starting up.

I've tested the packages on the following OS':

* Ubuntu 10.04, 12.04 14.04, 14.10 (64 bit) - deb installation
* CentOS 5 and 7 ( My 6 installation was buggy, can't install on VMWare ) (64 bit) - rpm installation
* Debian 5, 6 and 7 (64 bit) - deb installation
* CoreOS 653  - tgz installation

I've also added some utility methods to build the tarballs and deb / rpms, and ship them to packagecloud for apt / yum repo distribution. I will add a task to add a `build.md` file for more information about this.

Installation instructions are now in `INSTALL.md`, though they are pretty bare atm. I'll flesh out a bit more as I finish cleanup on this branch.

netshade commented:
Testing these packages by hand is pretty arduous btw. I feel like there's either a good 'scripting vagrant' or hijack some CI service to run a full integration test of installing a package, adding it to system startup, and running it, such that we have some guarantees on everything running okay.

janxious commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+### Unreleased
```
It's a marketing thing, but what if this was 1.0?

netshade replied:
No disagreement here on that change.

janxious replied:
1.0.0rc2 now, yah?

janxious commented on LICENSE:
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2011 Fastest Forward
+Copyright (c) 2015 Expected Behavior
```
one of us one of us

janxious commented on bin/instrument_server:
```diff
@@ -8,443 +8,211 @@ rescue Gem::LoadError
-        'filesystem.open_files_pct_max' => (open_files.to_f / max.to_f) * 100,
-      }
+if options[:user]
+  # Thanks Joe
```
?

netshade replied:
Durp, comment I wrote for myself to remember to thank the guy that wrote about [the correct way to drop privileges](http://timetobleed.com/5-things-you-dont-know-about-user-ids-that-will-destroy-you/). Will remove.

janxious commented on CUSTOM_METRICS.md:
```diff
@@ -0,0 +1,56 @@
+# Custom Metric Scripts
+
+You may have specific pieces of architecture that you would like `instrument_server` to monitor. As of version 0.7.0 of the `instrument_server` gem, you may pass the `-e` flag to `instrument_server` on startup to enable this functionality. There are several [examples](examples/) of scripts that you may use for your infrastructure, or you can [write your own](#writing_custom_scripts).
```
Update this version number once this goes out

netshade replied:
Definitely, will add a TODO in PR for this

janxious replied:
1.0.0rc2

janxious commented:
postgres probably wouldn't be the worst addition to the examples, and I wonder if mariadb/aurora would be easy to adapt for the mysql one.

janxious commented:
P.S. this is rad, man

netshade commented:
Yeah, definitely agreed re: Postgres. If I can I'd like to have something for all the infra we use in this gem, so we can take advantage of it as well.

And thanks! :)

janxious commented:
mongo, mysql, redis, postgres, memcache, (kinesis?,) pusher, (unicorn?,) nginx, kestrel, others?

netshade commented:
Cloudwatch would be pretty nice too, I think.

netshade commented:
Ready for re-review at this point. May be some more cleanup and some unit tests to add in here, but any future features I'd like to defer to a new release.

janxious commented on examples/docker/README.md:
```diff
@@ -0,0 +1,20 @@
+# Docker Metrics
```
What does it look like when you run instrument server on docker along with the docker script. Presumably one is pulling an incomprehensible hostname and one is using a human word set for the container name?

netshade replied:
Yeah, though the container names will be prefixed by the hostname. So if `worker-001` has Docker containers `doc-worker` and `job-runner` on it, you'll get Docker metrics that look like:

```
worker-001.docker.doc-worker.system.total
worker-001.docker.doc-worker.user_total
# etc
```

and

```
worker-001.docker.job-runner.system.total
worker-001.docker.job-runner.user_total
# etc
```

as well as `worker-001.cpu.in_use`, etc.

janxious commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+* Omit "-d" in favor of "start" and "stop", "foreground" runs process in foreground
+* Configurable reporting interval
+* Custom scripts may be executed and have their output sent to Instrumental (See [CUSTOM_METRICS.md](CUSTOM_METRICS.md))
+
```
I might add "docker support" to this specifically

janxious commented:
Version updates errrywhere, and :+1:

mediocretes commented on examples/mongo/mongo_3.rb:
```diff
@@ -0,0 +1,85 @@
+MONGO_AUTH_MECHANISM = ENV["MONGO_AUTH_MECHANISM"]
+MONGO_AUTH_DB        = ENV["MONGO_AUTH_DB"]
+
+CMDS                 = %w{mongotop mongostat}
```
Does this assume that mongotop/mongostat are in $PATH?  I don't know that that's true for everyone, many people use custom installs and scripts and ignore path entirely.  May need a configurable MONGO_PATH or something like that.

netshade replied:
Good point, will add

mediocretes commented on puppet/instrumental_tools/templates/instrumental.yml.erb:
```diff
@@ -0,0 +1,6 @@
+---
```
I would love to see more options available for use in here rather than as command line things, and then a more robust template... but that seems like a "next PR" sort of thing.

At some point though, as we monitor more kinds of things, we'll need more levers in here.

netshade replied:
Yeah, absolutely agreed. I could be swayed on changing this now, but this PR is so weighty already that yeah, I'd prefer to wait for next cycle.

mediocretes replied:
Are we introducing a lot of new command line options that we may want to remove once they are available via config file?  If so, it may make sense to wait.  If there aren't a lot of new options and/or we want to continue supporting them at the command line anyway, :shipit:

netshade replied:
We are introducing a fair number of new options in this release, but I don't think? we'll ever want to remove them. Like, I think the right philosophy to have here is that if a config file can specify it, it can also be specified via the command line.

netshade replied:
When I say "I think the right philosophy to have" there, you should imagine me shrugging my shoulders and sounding uncertain.

mediocretes replied:
Totally reasonable.

mediocretes replied:
Oh, I replied before the shrug thing.

In general, I like config files more than command lines and environments, but the things here are all single value default sort of options that are probably just fine at the command line.  Once we need to specify more complex arguments, it will be time to think about config file only options.  The defaults and simple arguments being accessible from the command line forever is probably a good thing.

TLDR: needs more discussion outside of this PR, ship it.

mediocretes commented on examples/mongo/README.md:
```diff
@@ -0,0 +1,25 @@
+# Mongo Metrics
+
+The [`mongo_3.rb`](mongo_3.rb) script generates metrics from the `mongotop` and `mongostat` commands for a Mongo 3.0 database. The following metrics will be output for every collection in your system:
```
Mongo 2 users... upgrade mongo if you want data?

A lot of this looks like it might work for mongo 2, not sure what demand is like for that.

netshade replied:
I'd actually be more pro Mongo 2, Mongo 3 support was an accident of me developing while having Mongo 3 locally, then realizing Mongo 2 tools don't support the JSON output mode. A Mongo 2 script would have a lot more text parsing, which is nbd, but would require doing the thing + testing.

mediocretes replied:
Cool.  Next PR then, and if anyone asks, "mongo 3 has better tools that made it easy" is a pretty reasonable answer.

mediocretes commented:
:+1: reviewed, did not run / test locally.

netshade commented:
:+1:
